Editor/EditorCore/MenuToggle.cs

```diff
@@ -0,0 +1,198 @@
+﻿
+#if UNITY_2019_1_OR_NEWER
+#define SHORTCUT_MANAGER
+#endif
+
+using System;
+using System.Diagnostics;
+using UnityEngine;
+using UnityEngine.ProBuilder;
+using Debug = UnityEngine.Debug;
+
+namespace UnityEditor.ProBuilder
+{
+    /// <summary>
+    /// Base class from which any action that is represented in the ProBuilder toolbar inherits.
+    /// </summary>
+    public abstract class MenuToggle: MenuAction
+    {
+        public enum MenuToggleState
+        {
+            Inactive,
+            Active,
+        };
+
+        MenuToggleState m_CurrentState;
+
+        protected delegate void StartEndCallBack();
+
+
+        protected MenuToggle()
+        {
+            iconMode = ProBuilderEditor.s_IsIconGui;
+            m_CurrentState = MenuToggleState.Inactive;
+
+            EditorApplication.update += OnUpdate;
+        }
+
+        public override void OnActionChanged(MenuAction action)
+        {
+            if(m_CurrentState == MenuToggleState.Active && !this.Equals(action))
+            {
+                Debug.Log("Changing menu action! Deactivating current toggle");
+                EndActivation(OnEnd);
+            }
+        }
+
+        /// <summary>
+        /// Not used for MenuToggle.
+        /// </summary>
+        /// <returns>A new ActionResult with a summary of the state of the action's success.</returns>
+        protected override ActionResult DoAction_Internal() {return ActionResult.Success;}
+
+        protected void OnStart()
+        {
+            m_CurrentState = MenuToggleState.Active;
+        }
+
+        protected void OnEnd()
+        {
+            m_CurrentState = MenuToggleState.Inactive;
+        }
+
+        /// <summary>
+        /// Perform whatever action this menu item is supposed to do when starting. You are responsible for implementing Undo.
+        /// </summary>
+        /// <returns>A new ActionResult with a summary of the state of the action's success.</returns>
+        protected abstract ActionResult StartActivation(StartEndCallBack callback);
+
+        /// <summary>
+        /// Call the Update for the current tool.
+        /// </summary>
+        public void OnUpdate()
+        {
+            if(m_CurrentState == MenuToggleState.Active)
+                UpdateAction();
+        }
```

Comment:
EditorTool manages it's own lifecycle. The Update/End action callbacks seem unnecessary, while Start could be put in the `DoAction` function

Reply:
Agree on the UpdateAction, this does not seems necessary, this was a first version of the menu toggle.
However, the problem with the end is still the same as a user can select another PB action in the menu without ending the tool first. As other actions are not EditorTools, do you have an idea on how to handle this?